run.c

```diff
@@ -331,21 +327,117 @@ Cell *copycell(Cell *x)	/* make a copy of a cell in a temp */
 {
 	Cell *y;
 
-	/* copy is not constant or field */
-
 	y = gettemp();
+	y->csub = CCOPY;
+
+	/* default to the uninitialized value */
+	if (x == NULL)
+		return y;
+
+	/* copy is not constant or field */
 	y->tval = x->tval & ~(CON|FLD|REC);
-	y->csub = CCOPY;	/* prevents freeing until call is over */
-	y->nval = x->nval;	/* BUG? */
-	if (isstr(x) /* || x->ctype == OCELL */) {
+
+	/*
+	 * What should we name this argument? The parser does not preserve
+	 * the names of parameters, so we don't know what the source calls
+	 * it. Is it worth the effort to change that?
+	 *
+	 * Note: Simply pointing to the incoming cell's name (as had been
+	 * done for a long time) creates a dangling pointer if that cell
+	 * is deleted (think array element).
+	 *
+	 * For now, args are anonymous. Debugging can use frame/arg indices.
+	 */
+	y->nval = NULL;
```

Comment:
The debugging statements need to be able to cope with anonymous cells. The temp cells used for expressions (including function return values) are anonymous. The jump cells are as well.

A quick grep found 13 instances of nval in DPRINTF statements, 10 wrapped, 3 unwrapped. I will wrap them all.

Also, parsing a function leaks the previous function definition's arglist. Compare `'function f(a) {}'` with `'function f(a) {} function g() {}'`. It will see about preserving the argument names while also eliminating the leak.

Reply:
It is not just DPRINTF calls, also SYNTAX, WARNING and FATAL.  Some of those may not be reachable but the ones in funnyvar() are.

Reply:
there is a lot to digest and review here. miguel, please keep NN wrapping as a separate PR. thanks

Reply:
@plan9, I will keep the NN wrapping separate. I hope to add a commit to this branch later today (perhaps tomorrow) to provide meaningful names for the args.